engines/director/lingo/xlibs/mmovie.cpp

```diff
@@ -126,32 +250,411 @@ void MMovieXObj::close(ObjectType type) {
+	g_lingo->push(Datum(0));
+}
+
 XOBJSTUB(MMovieXObj::m_getMovieNormalWidth, 0)
 XOBJSTUB(MMovieXObj::m_getMovieNormalHeight, 0)
-XOBJSTUB(MMovieXObj::m_getSegCount, 0)
-XOBJSTUB(MMovieXObj::m_getSegName, "")
-XOBJSTUB(MMovieXObj::m_getMovieRate, 0)
-XOBJSTUB(MMovieXObj::m_setMovieRate, 0)
+
+void MMovieXObj::m_getSegCount(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_getSegCount", nargs);
+	if (nargs != 1) {
+		g_lingo->dropStack(nargs);
+		g_lingo->push(MMovieError::MMOVIE_INVALID_MOVIE_INDEX);
+		return;
+	}
+	MMovieXObject *me = static_cast<MMovieXObject *>(g_lingo->_state->me.u.obj);
+	int movieIndex = g_lingo->pop().asInt();
+	if (!me->_movies.contains(movieIndex)) {
+		g_lingo->push(MMovieError::MMOVIE_INVALID_MOVIE_INDEX);
+		return;
+	}
+	g_lingo->push((int)me->_movies.getVal(movieIndex).segments.size());
+}
+
+void MMovieXObj::m_getSegName(int nargs) {
+	if (nargs != 2) {
+		g_lingo->dropStack(nargs);
+		g_lingo->push(Common::String(""));
+		return;
+	}
+	MMovieXObject *me = static_cast<MMovieXObject *>(g_lingo->_state->me.u.obj);
+	int segmentIndex = g_lingo->pop().asInt();
+	int movieIndex = g_lingo->pop().asInt();
+	if (!me->_movies.contains(movieIndex)) {
+		g_lingo->push(Common::String(""));
+		return;
+	}
+	if (segmentIndex > (int)me->_movies.getVal(movieIndex).segments.size() ||
+			segmentIndex <= 0) {
+		g_lingo->push(Common::String(""));
+		return;
+	}
+	Common::String result = me->_movies.getVal(movieIndex).segments[segmentIndex - 1]._name;
+	debugC(5, kDebugXObj, "MMovieXObj::m_getSegName(%d, %d): %s", movieIndex, segmentIndex, result.c_str());
+	g_lingo->push(result);
+}
+
+void MMovieXObj::m_getMovieRate(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_getMovieRate", nargs);
+	if (nargs != 0) {
+		g_lingo->dropStack(nargs);
+	}
+	MMovieXObject *me = static_cast<MMovieXObject *>(g_lingo->_state->me.u.obj);
+	g_lingo->push(Datum(me->_rate));
+}
+
+void MMovieXObj::m_setMovieRate(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_setMovieRate", nargs);
+	if (nargs != 1) {
+		warning("MMovieXObj::m_setMovieRate: expecting 4 arguments");
+		g_lingo->dropStack(nargs);
+		g_lingo->push(Datum(0));
+		return;
+	}
+	MMovieXObject *me = static_cast<MMovieXObject *>(g_lingo->_state->me.u.obj);
+	me->_rate = g_lingo->pop().asInt();
+	g_lingo->push(Datum(me->_rate));
+}
+
 XOBJSTUB(MMovieXObj::m_flushEvents, 0)
 XOBJSTUB(MMovieXObj::m_invalidateRect, 0)
-XOBJSTUB(MMovieXObj::m_readFile, "")
-XOBJSTUB(MMovieXObj::m_writeFile, "")
+
+void MMovieXObj::m_readFile(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_readFile", nargs);
+	if (nargs != 2) {
+		warning("MMovieXObj::m_readFile(): expecting 2 argument");
+	}
+	Common::SaveFileManager *saves = g_system->getSavefileManager();
+	bool scramble = g_lingo->pop().asInt() != 0;
+	Common::String origPath = g_lingo->pop().asString();
+	Common::String path = origPath;
+
+	Common::String prefix = g_director->getTargetName() + '-';
```

Comment:
I think, we need to factor this code out to utils.cpp, since it is also used by at least FileIO Xtra

Reply:
Done

Reply:
Hmmm. I also meant something like:
```C++
Common::String generateSaveFilePrefix() {
   return g_director->getTargetName() + '-';
 }
 ```

Reply:
Done